content/concepts/consensus-network/invariant-checking.md

```diff
@@ -0,0 +1,404 @@
+
+The invariant checker is then invariantly checked to ensure the invariant checker's failure is not due to a bug. If this second level check fails, transactions are marked with a **tefINVARIANT_FAILED** result code. The transaction is in its final state and is not included in the ledger.
+
+
+## Active Invariant Checks
+
+Following is a list of active checks that the invariant checker runs against each transaction on the XRP Ledger.
+
+- [Transaction Fee Check](#transaction-fee-check)
+- [XRP Not Created](#xrp-not-created)
+- [Account Roots Not Deleted](#account-roots-not-deleted)
+- [XRP Balance Checks](#xrp-balance-checks)
+- [Ledger Entry Types Match](#ledger-entry-types-match)
+- [No XRP Trust Lines](#no-xrp-trust-lines)
+- [No Bad Offers](#no-bad-offers)
+- [No Zero Escrow](#no-zero-escrow)
+- [Valid New Account Root](#valid-new-account-root)
+
+
+### Transaction Fee Check
+
+- **Condition(s) Checked:**
+
+    - [Trnsaction fees](https://xrpl.org/rippleapi-reference.html#transaction-fees) should never be negative nor larger than the transaction itself.
+
+```
+class TransactionFeeCheck
+{
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
+        STTx const&,
+        TER const,
+        XRPAmount const,
+        ReadView const&,
+        beast::Journal const&);
+};
+```
+
+### XRP Not Created
+
+- **Condition(s) Checked:**
+    
+    - A transaction must not create XRP and should only destroy the XRP [fee](https://xrpl.org/rippleapi-reference.html#transaction-fees)).
+
+```
+class XRPNotCreated
+{
+    std::int64_t drops_ = 0;
+
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
+        STTx const&,
+        TER const,
+        XRPAmount const,
+        ReadView const&,
+        beast::Journal const&);
+};
+```
+
+### Account Roots Not Deleted
+
+- **Condition(s) Checked:**
+    
+    - An account ledger entry cannot be removed.
+
+```
+class AccountRootsNotDeleted
+{
+    std::uint32_t accountsDeleted_ = 0;
+
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
+        STTx const&,
+        TER const,
+        XRPAmount const,
+        ReadView const&,
+        beast::Journal const&);
+};
+```
+
+### XRP Balance Checks
+
+- **Condition(s) Checked:**
+    
+    - An account's XRP balance must be of type XRP and take a value between 0 and INITIAL_XRP drops, inclusive.
```

Comment:
Instead "and take a value between 0 and INITIAL_XRP drops, inclusive" we should write something an ordinary person could understand without more background knowledge. 

```suggestion
    - An account's XRP balance must be of type XRP, and it cannot be less than 0 or more than 100 billion XRP exactly.
```